File: src/research_agent/eval/metrics.py
"""Pure evaluation metrics: no I/O, no network, no models."""

import math
from itertools import product

from ..jev import JevThresholds, decide_from_probabilities

Z95 = 1.959964


def wilson(k, n, z=Z95):
    """Wilson score interval for k successes in n trials; None when n == 0."""
    if n == 0:
        return None
    p = k / n
    denom = 1 + z * z / n
    centre = (p + z * z / (2 * n)) / denom
    half = z * math.sqrt(p * (1 - p) / n + z * z / (4 * n * n)) / denom
    return (max(0.0, centre - half), min(1.0, centre + half))


def rate(k, n):
    """k of n with a Wilson interval. A zero denominator gives value None plus a reason, never 100%."""
    if n == 0:
        return {"k": k, "n": n, "value": None, "ci": None, "reason": "zero denominator"}
    return {"k": k, "n": n, "value": k / n, "ci": wilson(k, n), "reason": None}


def _check_pairs(a, b):
    if len(a) != len(b) or not a:
        raise ValueError("kappa needs two equal-length, non-empty label lists")


def cohen_kappa(a, b):
    """Cohen's kappa for two raters. Reports raw agreement and prevalence, because kappa collapses
    when one class dominates. Single-class input gives kappa None plus a reason."""
    _check_pairs(a, b)
    n = len(a)
    labels = sorted(set(a) | set(b), key=str)
    observed = sum(x == y for x, y in zip(a, b, strict=True)) / n
    expected = sum((a.count(label) / n) * (b.count(label) / n) for label in labels)
    result = {
        "n": n,
        "agreement": observed,
        "prevalence": {label: (a.count(label) + b.count(label)) / (2 * n) for label in labels},
    }
    if expected == 1:
        return {**result, "kappa": None, "reason": "single class: kappa undefined"}
    return {**result, "kappa": (observed - expected) / (1 - expected), "reason": None}


def weighted_kappa(a, b, levels=range(5)):
    """Quadratic-weighted kappa for ordinal scores (default 0..4)."""
    _check_pairs(a, b)
    levels = list(levels)
    k = len(levels)
    if k < 2:
        raise ValueError("weighted kappa needs at least two levels")
    if not set(a) <= set(levels) or not set(b) <= set(levels):
        raise ValueError(f"scores must all be in levels {levels}")
    index = {value: i for i, value in enumerate(levels)}
    n = len(a)
    observed = [[0.0] * k for _ in range(k)]
    for x, y in zip(a, b, strict=True):
        observed[index[x]][index[y]] += 1 / n
    rows = [sum(r) for r in observed]
    cols = [sum(observed[i][j] for i in range(k)) for j in range(k)]
    num = den = 0.0
    for i, j in product(range(k), repeat=2):
        weight = ((i - j) / (k - 1)) ** 2
        num += weight * observed[i][j]
        den += weight * rows[i] * cols[j]
    result = {
        "n": n,
        "agreement": sum(x == y for x, y in zip(a, b, strict=True)) / n,
        "prevalence": {
            level: (a.count(level) + b.count(level)) / (2 * n) for level in levels if level in a or level in b
        },
    }
    if den == 0:
        return {**result, "kappa": None, "reason": "single class: kappa undefined"}
    return {**result, "kappa": 1 - num / den, "reason": None}


INCLUDE_GRID = [round(i / 10, 1) for i in range(1, 10)]
EXCLUDE_GRID = [0.5, 0.6, 0.7, 0.8, 0.9, 0.95, 0.99]
STRATEGIES = ("llm_only", "jev_only", "cascade")


def cascade_decision(probabilities, llm_decision, thresholds):
    """The shipped cascade: a confident Jev verdict decides, otherwise the LLM screen does."""
    verdict = decide_from_probabilities(probabilities, thresholds)
    return (verdict, "jev") if verdict != "escalate" else (llm_decision, "llm")


def _decide(record, strategy, thresholds):
    if strategy == "llm_only":
        return record["llm"], "llm"
    if strategy == "jev_only":
        return decide_from_probabilities(record["probabilities"], thresholds), "jev"
    if strategy == "cascade":
        return cascade_decision(record["probabilities"], record["llm"], thresholds)
    raise ValueError(f"unknown strategy {strategy!r}")


def evaluate(records, strategy, thresholds):
    """Recall on SR-included papers plus workload. 'Kept' means not excluded, as in the pipeline."""
    decided = [(r, *_decide(r, strategy, thresholds)) for r in records]
    positives = [r for r in records if r["label"] == "include"]
    missed = [
        {"id": r["id"], "title": r["title"], "probabilities": r["probabilities"], "decision": d, "tier": t}
        for r, d, t in decided
        if r["label"] == "include" and d == "exclude"
    ]
    # Positives this strategy excludes although llm_only would keep them: the regression the pipeline must avoid.
    lost = (
        []
        if strategy == "llm_only"
        else [
            {"id": r["id"], "title": r["title"], "probabilities": r["probabilities"], "tier": t}
            for r, d, t in decided
            if r["label"] == "include" and d == "exclude" and r["llm"] != "exclude"
        ]
    )
    n = len(records)
    kept = [r for r, d, _t in decided if d != "exclude"]
    if strategy == "llm_only":
        auto_include = auto_exclude = 0
        escalated, calls_saved = n, 0
    else:
        verdicts = [decide_from_probabilities(r["probabilities"], thresholds) for r in records]
        auto_include, auto_exclude = verdicts.count("include"), verdicts.count("exclude")
        escalated = verdicts.count("escalate")
        calls_saved = n if strategy == "jev_only" else n - escalated
    return {
        "recall": rate(len(positives) - len(missed), len(positives)),
        "missed": missed,
        "lost_vs_llm": lost,
        "auto_include": auto_include,
        "auto_exclude": auto_exclude,
        "escalated": escalated,
        "calls_saved": calls_saved,
        # Downstream cost: kept papers go on to extraction and both reviewers.
        "kept": len(kept),
        "kept_negatives": sum(r["label"] == "not_included" for r in kept),
    }


def sweep(records, includes=INCLUDE_GRID, excludes=EXCLUDE_GRID):
    """Cascade outcome for every allowed threshold pair (exclude must be >= include: recall first)."""
    rows = []
    for include, exclude in product(includes, excludes):
        if exclude < include:
            continue
        out = evaluate(records, "cascade", JevThresholds(include, exclude))
        rows.append(
            {
                "min_confidence": include,
                "exclude_min_confidence": exclude,
                "recall": out["recall"],
                "missed": len(out["missed"]),
                "lost_vs_llm": len(out["lost_vs_llm"]),
                "lost_ids": [m["id"] for m in out["lost_vs_llm"]],
                "calls_saved": out["calls_saved"],
                "auto_include": out["auto_include"],
                "auto_exclude": out["auto_exclude"],
                "escalated": out["escalated"],
                "kept": out["kept"],
                "kept_negatives": out["kept_negatives"],
            }
        )
    return rows


def recommend(rows, target=None, other_rows=None):
    """Admissible = loses no SR-included paper that llm_only keeps (`lost_vs_llm == 0`) on `rows` and, when
    given, on `other_rows` (the same threshold pair; a pair missing there is not admissible), and, only
    when `target` is set, point-estimate recall >= target on `rows`. Rank: most calls saved (summed over
    both row sets), then fewer missed, then the stricter include, then the stricter exclude threshold.
    None when nothing is admissible: never silently pick the 'least bad' pair."""
    other = None
    if other_rows is not None:
        other = {(r["min_confidence"], r["exclude_min_confidence"]): r for r in other_rows}
    candidates = []
    for r in rows:
        if r["lost_vs_llm"] != 0:
            continue
        if target is not None and (r["recall"]["value"] is None or r["recall"]["value"] < target):
            continue
        partner = None
        if other is not None:
            partner = other.get((r["min_confidence"], r["exclude_min_confidence"]))
            if partner is None or partner["lost_vs_llm"] != 0:
                continue
        saved = r["calls_saved"] + (partner["calls_saved"] if partner else 0)
        missed = r["missed"] + (partner["missed"] if partner else 0)
        candidates.append(((-saved, missed, -r["min_confidence"], -r["exclude_min_confidence"]), r))
    return min(candidates, key=lambda c: c[0])[1] if candidates else None

File: web/src/features/evals/metrics.ts
export type Rate = { k: number; n: number; value: number | null; ci: [number, number] | null; reason: string | null };
export type Pair = { include: number; exclude: number };
export type SweepRow = { pair: Pair; recall: Rate | null; missed: number; lost: number; lostIds: string[]; callsSaved: number; escalated: number };
export type Strategy = { name: "llm_only" | "jev_only" | "cascade"; recall: Rate | null; callsSaved: number; escalated: number; kept: number; keptNegatives: number };
export type EvalView = {
  gold: { name: string; citation: string };
  screened: number | null;
  retrievalRecall: Rate | null;
  defaultPair: Pair | null;
  strategies: Strategy[];
  sweep: SweepRow[];
  holdout: { gold: string; n: number; recall: Rate | null; pair: Pair | null; callsSaved: number | null } | null;
  holdoutSweep: { gold: string; n: number; rows: SweepRow[] } | null;
  recommended: Pair | null;
  rejected: { pair: Pair; lost: { id: string; title: string }[] } | null;
  agreement: { n: number; kappa: number | null; agreement: number | null; reason: string | null; sameFamily: boolean | null; adjudicated: Rate | null } | null;
  warnings: string[];
};

type Json = Record<string, unknown>;
const obj = (value: unknown): Json | null => (value && typeof value === "object" && !Array.isArray(value) ? (value as Json) : null);
const num = (value: unknown): number | null => (typeof value === "number" && Number.isFinite(value) ? value : null);
const str = (value: unknown, fallback = ""): string => (typeof value === "string" ? value : fallback);

export function toRate(value: unknown): Rate | null {
  const r = obj(value);
  if (!r || num(r.k) === null || num(r.n) === null) return null;
  const ci = Array.isArray(r.ci) && r.ci.length === 2 && r.ci.every((x) => num(x) !== null) ? ([r.ci[0], r.ci[1]] as [number, number]) : null;
  return { k: r.k as number, n: r.n as number, value: num(r.value), ci, reason: typeof r.reason === "string" ? r.reason : null };
}

const toPair = (value: unknown): Pair | null => {
  const p = obj(value);
  const include = num(p?.min_confidence);
  const exclude = num(p?.exclude_min_confidence);
  return include === null || exclude === null ? null : { include, exclude };
};

function toRows(value: unknown): SweepRow[] {
  if (!Array.isArray(value)) return [];
  return value.flatMap((item) => {
    const row = obj(item);
    const pair = toPair(row);
    if (!row || !pair) return [];
    return [{
      pair, recall: toRate(row.recall), missed: num(row.missed) ?? 0, lost: num(row.lost_vs_llm) ?? 0,
      lostIds: Array.isArray(row.lost_ids) ? row.lost_ids.filter((x): x is string => typeof x === "string") : [],
      callsSaved: num(row.calls_saved) ?? 0, escalated: num(row.escalated) ?? 0,
    }];
  });
}

export function parseEval(metrics: Json): EvalView {
  const gold = obj(metrics.gold);
  const strategies = obj(metrics.strategies) ?? {};
  const holdout = obj(metrics.holdout);
  const holdoutSweep = obj(metrics.holdout_sweep);
  const rejected = obj(metrics.rejected_on_holdout);
  const agreement = obj(metrics.agreement);
  const verdict = obj(agreement?.verdict);
  return {
    gold: { name: str(gold?.name, "unknown"), citation: str(gold?.citation) },
    screened: num(obj(metrics.counts)?.screened),
    retrievalRecall: toRate(metrics.retrieval_recall),
    defaultPair: toPair(metrics.default_thresholds),
    strategies: (["llm_only", "jev_only", "cascade"] as const).flatMap((name) => {
      const s = obj(strategies[name]);
      return s ? [{ name, recall: toRate(s.recall), callsSaved: num(s.calls_saved) ?? 0, escalated: num(s.escalated) ?? 0, kept: num(s.kept) ?? 0, keptNegatives: num(s.kept_negatives) ?? 0 }] : [];
    }),
    sweep: toRows(metrics.sweep),
    holdout: holdout ? { gold: str(holdout.gold), n: num(holdout.n) ?? 0, recall: toRate(holdout.recall), pair: toPair(holdout.thresholds), callsSaved: num(holdout.calls_saved) } : null,
    holdoutSweep: holdoutSweep ? { gold: str(holdoutSweep.gold), n: num(holdoutSweep.n) ?? 0, rows: toRows(holdoutSweep.rows) } : null,
    recommended: toPair(metrics.recommended),
    rejected: rejected && toPair(rejected.thresholds)
      ? {
          pair: toPair(rejected.thresholds)!,
          lost: Array.isArray(rejected.lost) ? rejected.lost.flatMap((m) => { const o = obj(m); return o ? [{ id: str(o.id), title: str(o.title) }] : []; }) : [],
        }
      : null,
    agreement: agreement
      ? { n: num(agreement.n) ?? 0, kappa: num(verdict?.kappa), agreement: num(verdict?.agreement), reason: typeof verdict?.reason === "string" ? verdict.reason : null, sameFamily: typeof agreement.same_family === "boolean" ? agreement.same_family : null, adjudicated: toRate(agreement.adjudication_rate) }
      : null,
    warnings: Array.isArray(metrics.warnings) ? metrics.warnings.filter((w): w is string => typeof w === "string") : [],
  };
}

export function formatRate(rate: Rate | null): string {
  if (!rate) return "n/a";
  if (rate.value === null || !rate.ci) return `n/a (${rate.reason ?? "undefined"})`;
  return `${rate.k}/${rate.n} (95% CI ${rate.ci[0].toFixed(2)}–${rate.ci[1].toFixed(2)})`;
}

export const rowFor = (rows: SweepRow[], pair: Pair): SweepRow | undefined =>
  rows.find((row) => row.pair.include === pair.include && row.pair.exclude === pair.exclude);

export const samePair = (a: Pair | null, b: Pair | null) => !!a && !!b && a.include === b.include && a.exclude === b.exclude;
